crates/accelerate/src/synthesis/linear/lnn.rs

```diff
@@ -0,0 +1,217 @@
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use std::iter::once;
+
+use hashbrown::HashMap;
+use itertools::Itertools;
+use ndarray::{Array1, ArrayView2};
+
+use qiskit_circuit::{
+    operations::{Param, StandardGate},
+    Qubit,
+};
+use smallvec::{smallvec, SmallVec};
+
+// A sequence of Lnn gates
+// Represents the return type for Lnn Synthesis algorithms
+type LnnGatesVec = Vec<(StandardGate, SmallVec<[Param; 3]>, SmallVec<[Qubit; 2]>)>;
+
+// A pattern denoted by Pj in [1] for odd number of qubits:
+// [n-2, n-4, n-4, ..., 3, 3, 1, 1, 0, 0, 2, 2, ..., n-3, n-3]
+fn _odd_pattern1(n: isize) -> Vec<isize> {
+    once(n - 2)
+        .chain((0..((n - 3) / 2)).flat_map(|i| [(n - 2 * i - 4); 2]))
+        .chain((0..((n - 1) / 2)).flat_map(|i| [2 * i; 2]))
+        .collect()
+}
+
+// A pattern denoted by Pk in [1] for odd number of qubits:
+// [2, 2, 4, 4, ..., n-1, n-1, n-2, n-2, n-4, n-4, ..., 5, 5, 3, 3, 1]
+fn _odd_pattern2(n: isize) -> Vec<isize> {
+    (0..((n - 1) / 2))
+        .flat_map(|i| [(2 * i + 2); 2])
+        .chain((0..((n - 3) / 2)).flat_map(|i| [n - 2 * i - 2; 2]))
+        .chain(once(1))
+        .collect()
+}
+
+// A pattern denoted by Pj in [1] for even number of qubits:
+// [n-1, n-3, n-3, n-5, n-5, ..., 1, 1, 0, 0, 2, 2, ..., n-4, n-4, n-2]
+fn _even_pattern1(n: isize) -> Vec<isize> {
+    once(n - 1)
+        .chain((0..((n - 2) / 2)).flat_map(|i| [n - 2 * i - 3; 2]))
+        .chain((0..((n - 2) / 2)).flat_map(|i| [2 * i; 2]))
+        .chain(once(n - 2))
+        .collect()
+}
+
+// A pattern denoted by Pk in [1] for even number of qubits:
+// [2, 2, 4, 4, ..., n-2, n-2, n-1, n-1, ..., 3, 3, 1, 1]
+fn _even_pattern2(n: isize) -> Vec<isize> {
+    (0..((n - 2) / 2))
+        .flat_map(|i| [2 * (i + 1); 2])
+        .chain((0..(n / 2)).flat_map(|i| [(n - 2 * i - 1); 2]))
+        .collect()
+}
+
+// Creating the patterns for the phase layers.
+fn _create_patterns(n: isize) -> HashMap<(isize, isize), (isize, isize)> {
+    let (pat1, pat2) = if n % 2 == 0 {
+        (_even_pattern1(n), _even_pattern2(n))
+    } else {
+        (_odd_pattern1(n), _odd_pattern2(n))
+    };
+
+    let ind = if n % 2 == 0 {
+        (2 * n - 4) / 2
+    } else {
+        (2 * n - 4) / 2 - 1
+    };
+
+    HashMap::from_iter((0..n).map(|i| ((0, i), (i, i))).chain(
+        (0..(n / 2)).cartesian_product(0..n).map(|(layer, i)| {
+            (
+                (layer + 1, i),
+                (
+                    pat1[(ind - (2 * layer) + i) as usize],
+                    pat2[((2 * layer) + i) as usize],
+                ),
+            )
+        }),
+    ))
+}
+
+// A single layer of CX gates.
+fn _append_cx_stage1(gates: &mut LnnGatesVec, n: usize) {
```

Comment:
I think that this function  `_append_cx_stage1` as well as  `_append_cx_stage2` should belong to 
`crates/accelerate/src/synthesis/permutation`
since it's part of the [reverse permutation code](https://github.com/Qiskit/qiskit/blob/main/qiskit/synthesis/permutation/permutation_reverse_lnn.py).

Could you please also port `_append_reverse_permutation_lnn_kms` to rust ?